### PR TITLE
Simplify TemplateSource, CombinedSource and SpectrumTemplateSource

### PR DESCRIPTION
**Pull request by dachengx**
## What does the code in this PR do / what does it improve?

The functionality of `TemplateSource`, `CombinedSource`, and `SpectrumTemplateSource` are unchanged.

Add test of `CombinedSource`. 

The weights can be set as shape parameters in the config. You can check it by:

```
config = load_yaml('unbinned_wimp_statistical_model.yaml')
blueice_model = BlueiceExtendedModel(
    parameter_definition=config['parameter_definition'],
    likelihood_config=config['likelihood_config'],
)

llh_term = blueice_model.get_likelihood_term_from_name('sr2')
zs = np.asarray([-2, 0])

blueice_model._likelihood.likelihood_list[llh_term].mus_interpolator(zs)
blueice_model._likelihood.likelihood_list[llh_term].ps_interpolator(zs).std()
```

But the `SpectrumTemplateSource` is not tested, also the configuration is too complex, especially the normalization. 

## Can you briefly describe how it works?

I already tested that the results are the same in `main` and `simplify_source`. 

## Can you give a minimal working example (or illustrate with a figure)?

Please refer to `test_blueice_extended_model.py`. 

## What are the potential drawbacks of the codes?

_Please include the following if applicable:_
  - [x] _Update the docstring(s)_
  - [x] _Update the documentation_
  - [x] _Tests to check the (new) code is working as desired._
  - [ ] _Does it solve one of the open issues on github?_

### CODE REVIEW

**Comment by dachengx** [comment]
Maybe we also want to test `SpectrumTemplateSource`.

**Comment by dachengx** [comment]
This PR depends on https://github.com/XENONnT/alea/pull/68.

**Review by kdund — CHANGES_REQUESTED** [review]
SOme comments-- at the moment, the only way I found testing it was to import binference plotting code, is there a better way, you think?

File: alea/examples/configs/unbinned_wimp_statistical_model.yaml [hunk 1]
```diff
@@ -127,3 +132,38 @@ likelihood_config:
...
+      sources:
+      - name: er
+        class: alea.template_source.CombinedSource
+        parameters:
+          - er_rate_multiplier
+          - er_band_shift
+        named_parameters:
+          - er_band_shift
+        fixed_weight: 0.2
+        weight_names: [fixed_weight, er_band_shift]  # not meaningful, just an example
```
[the remark this answers is not in the document]
Comment (dachengx) [reply alea/examples/configs/unbinned_wimp_statistical_model.yaml:155]: To show that we can use fixed weight and also shape parameters. Also to make it harder for the test.

File: alea/examples/configs/unbinned_wimp_statistical_model.yaml [hunk 2]
```diff
@@ -127,3 +132,38 @@ likelihood_config:
         template_filename: wimp50gev_template.h5
         apply_efficiency: True
         efficiency_name: signal_efficiency
+
+    # SR2
+    - name: sr2
+      default_source_class: alea.template_source.TemplateSource
+      likelihood_type: blueice.likelihood.UnbinnedLogLikelihood
+      analysis_space:
+        - "cs1": 'np.arange(0, 102, 2)'
```
[the remark this answers is not in the document]
Comment (dachengx) [reply alea/examples/configs/unbinned_wimp_statistical_model.yaml:141]: No objections. @hammannr Do you have any opinions?

File: alea/template_source.py [hunk 3]
```diff
@@ -66,174 +151,138 @@ def build_histogram(self):
...
+        self.events_per_day = h.n
         logging.debug(f"events_per_day: " + str(self.events_per_day))
 
+        # Convert h to density...
+        if self.config.get("in_events_per_bin", True):
+            h.histogram /= self._bin_volumes
+
         # ... and finally to probability density
-        if 0 < self.events_per_day:
+        if self.events_per_day > 0:
```
[the remark this answers is not in the document]
Comment (dachengx) [reply alea/template_source.py:197]: We will apply `a < b < c` in the future.

File: alea/template_source.py [hunk 4]
```diff
@@ -66,174 +151,138 @@ def build_histogram(self):
...
-        for i, (n, _) in enumerate(self.config.get('analysis_space')):
+        for i, (n, _) in enumerate(self.config["analysis_space"]):
             ret[n] = t[:, i]
         return ret
 
 
-class CombinedSource(blueice.HistogramPdfSource):
+class CombinedSource(TemplateSource, HistogramPdfSource):
     """
     Source that inherits structure from TH2DSource by Jelle,
```
[the remark this answers is not in the document]
Comment (dachengx) [reply alea/template_source.py:229]: removed.

**Comment by dachengx** [comment]
I think `histogram_multiplier` is a bad name, actually it is the absolute rate.

File: alea/examples/configs/unbinned_wimp_statistical_model.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (hammannr) [reply alea/examples/configs/unbinned_wimp_statistical_model.yaml:141]: I think I prefer np.arange but I don't mind linspace in the example file.

**Comment by hammannr** [comment]
> I think `histogram_multiplier` is a bad name, actually it is the absolute rate.

It depends on how you use it and I think we agreed that we try to avoid people using it. But really, the histogram is just multiplied with it, right?

File: alea/parameters.py [hunk 5]
```diff
@@ -1,5 +1,7 @@
 from typing import Any, Dict, List, Optional, Tuple
 
+import scipy
```
[the remark this answers is not in the document]
Comment (hammannr) [reply alea/parameters.py:3]: Why do you import it though @dachengx ? If there's a reason you can also just put the #noqa: F401 to make clear that it is intentionally added here and nobody accidentally removes it in the future 😊

File: alea/template_source.py [hunk 6]
```diff
@@ -2,62 +2,147 @@
...
 import logging
 
-import blueice
 import numpy as np
 from scipy.interpolate import interp1d
+from blueice import HistogramPdfSource
 from inference_interface import template_to_multihist
 
-logging.basicConfig(level=logging.INFO)
+logging.basicConfig(level=logging.debug)
```
Comment (hammannr) [inline alea/template_source.py:10]: Ok for testing but we should revert this before merging

**Review by hammannr — COMMENTED** [review]
I scrolled through the changes and it looks nice, for sure useful to clean up this part. Though a bit hard to keep track and check all the changes. Maybe it would make sense to add a test for the combined source or did I just miss it?
Also, could you maybe think of a case where this source would be used so that we can make a sensible example out of it?

File: alea/examples/configs/unbinned_wimp_statistical_model.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (kdund) [reply alea/examples/configs/unbinned_wimp_statistical_model.yaml:141]: np.arange hides the endpoint (100) from the unwary

File: alea/parameters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (dachengx) [reply alea/parameters.py]: Added.

**Comment by dachengx** [comment]
> > I think `histogram_multiplier` is a bad name, actually it is the absolute rate.
> 
> It depends on how you use it and I think we agreed that we try to avoid people using it. But really, the histogram is just multiplied with it, right?

First, the histogram is normalized to one and then multiplied by it.

File: alea/template_source.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dachengx) [reply alea/template_source.py]: Removed.

File: alea/template_source.py (hunk 6 above)
Comment (dachengx) [reply alea/template_source.py]: Reverted.

**Comment by dachengx** [comment]
`histogram_scale_factor` is also multiplied on the histogram.

**Comment by hammannr** [comment]
Ah, right @dachengx ! The `histogram_multiplier` was renamed to `histogram_scale_factor` in the `TemplateSource`. For the `CombinedSource` it indeed is the rate before slicing the way it is implemented. I never worked with the CombinedSource, do we need/want this behaviour? Otherwise we might consider sticking to a `histogram_scale_factor` for both that do the same thing i.e. just being multiplied to the template. If you want to normalize things to one you should do so yourself. But in any case, the intended usage is to only use the multiplier if you very quickly want to check or vary something. For production you should update the normalization of the template according to the expectation value, at least this is what we discussed with @kdund I think.

**Comment by hammannr** [comment]
Ok I see that for now there is no urgency to have a meaningful test but it would be very useful to have an idea when this source could be used and how one would want to use it, I think. Maybe we could also make this a bit clearer in the docstring since rn I have no idea what to do with it tbh 😄

**Comment by dachengx** [comment]
For this to be merged, we need more example templates like https://github.com/XENONnT/alea/pull/30 provided. Especially the 3D template with `(E, cS2, cS2)`.

**Comment by dachengx** [comment]
@kdund I moved the test of sources into another test module. They will not participate in CL calculation, to save time.

File: README.md [hunk 7]
```diff
@@ -27,8 +27,8 @@ You are now ready to use alea!
 
 ## Getting started
 The best way to get started is to check out the [documentation](https://alea.readthedocs.io/en/latest/) and have a look at our [tutorial notebooks](https://github.com/XENONnT/alea/tree/main/notebooks). To explore the notebooks interactively, you can use [Binder](https://mybinder.org/v2/gh/XENONnT/alea/HEAD?labpath=notebooks).
-## Ackgnowledgements
 
+## Ackgnowledgements
```
Comment (kdund) [inline README.md]: acknowledgements :)

**Review by kdund — APPROVED** [review]
Thanks, @dachengx I think from my side this looks good now

**Pull request merged**